addons/mail/controllers/mail.py

```diff
@@ -165,3 +165,26 @@ def mail_action_view(self, model=None, res_id=None, access_token=None, **kwargs)
             except ValueError:
                 res_id = False
         return self._redirect_to_record(model, res_id, access_token, **kwargs)
+
+    @http.route('/mail/unfollow', type='http', auth='public')
+    def mail_action_unfollow(self, model, res_id, pid, token, **kwargs):
+        comparison, record, redirect = MailController._check_token_and_record_or_redirect(model, int(res_id), token)
+        if not comparison or not record:
+            raise AccessError(_('Non existing record or wrong token.'))
+
+        pid = int(pid)
+        record_sudo = record.sudo()
+        if pid in record_sudo.message_partner_ids.ids:
+            partner = request.env['res.partner'].sudo().browse([pid])
+            record_sudo.message_unsubscribe([pid])
+            record_sudo._message_log(body=_('%(user_name)s has unfollowed the document.', user_name=partner.name))
+
+        if request.session.uid:
+            try:
+                record.check_access_rights('read')
+                record.check_access_rule('read')
+            except AccessError:
+                pass
```

Comment:
what's the purpose of this ?

Reply:
If the user is logged and has read access to the record, redirect the user to the record otherwise redirect it to a simple page indicating that the user don’t follow the record anymore. 

I am not using the “redirect” from _check_token_and_record_or_redirect because it redirect to messaging in case the record cannot be read by the user instead of the feeback page indicating that the user don’t follow the document anymore.

When the user has access to the record, he is redirected to the record and a message is added in the chatter that says that he doesn’t follow the document anymore (a toast was not possible without a custom development).

Reply:
woops sorry, as discussed on discord dumb STD can't read an indentation